### PR TITLE
[Serializer] Refactor and uniformize the config by introducing a default context

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no     
| Deprecations? | yes 
| Tests pass?   | yes    
| Fixed tickets | n/a   
| License       | MIT
| Doc PR        | todo 

This PR uniformizes how the Serializer's configuration is handled:

* As currently, configuration options can be set using the context (options that weren't configurable using the context have been refactored to leverage it)
* Normalizers and encoders' constructors now accept a "default context"
* All existing global configuration flags (constructor parameters) have been deprecated in favor of this new default context
* the stateless context is always tried first, then is the default context

Some examples:

```php
// Configuring groups globally
// Before: not possible
// After
$normalizer = new ObjectNormalizer(/* deps */, ['groups' => 'the_default_group']);

// Escaping Excel-like formulas in CSV files
// Before
$encoder = new CsvEncoder(',', '"', '\\', '.', true);
// After
$encoder = new CsvEncoder(['csv_escape_formulas' => true]);
$encoder->normalize($data, 'csv', ['csv_escape_formulas' => false]); // Override for this call only
```

Benefits:

* The DX is dramatically improved, configuration is always handled in similar way
* The serializer can be used in fully stateless way
* Every options can be configured globally
* Classes that had constructors with a lot of parameters (like `CsvEncoder`) are now much easier to use
* We'll be able to improve the documentation by adding a dictionary of all available context options for the whole component
* Everything can be configured the same way

TODO in subsequent PRs:

* Add a new option in framework bundle to configure the context globally
* Uniformize the constants name (sometimes the name if `FOO`, sometimes `FOO_KEY`)
* Fix the "bug" regarding the format configuration in `DateTimeNormalizer::denormalize()`  (see comments)
* Maybe: move `$defaultContext` as the first parameter (before required services?)
* Make `XmlEncoder` stateless

### CODE REVIEW

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php [hunk 1]
```diff
@@ -76,7 +94,7 @@ public function supportsNormalization($data, $format = null)
      */
     public function denormalize($data, $class, $format = null, array $context = array())
     {
-        $dateTimeFormat = isset($context[self::FORMAT_KEY]) ? $context[self::FORMAT_KEY] : null;
+        $dateTimeFormat = $context[static::FORMAT_KEY] ?? null;
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: Even before this PR, `$this->format` was ignored when denormalizing (but not when normalizing).
Changing this behavior introduces a BC break, and breaks the tests.

An alternative could be to introduce a new key, like `denormalize_datetime_format`, and to deprecate using `datetime_format` when denormalizing.

Anyway, it's out of this PR scope.

File: src/Symfony/Component/Serializer/Encoder/CsvEncoder.php [hunk 2]
```diff
@@ -30,20 +30,45 @@ class CsvEncoder implements EncoderInterface, DecoderInterface
...
+            array('enclosure', static::ENCLOSURE_KEY, '"'),
+            array('escapeChar', static::ESCAPE_CHAR_KEY,  '\\'),
+            array('keySeparator', static::KEY_SEPARATOR_KEY, '.'),
+            array('escapeFormulas', static::ESCAPE_FORMULAS_KEY, false),
+        );
+        foreach ($args as $arg) {
+            $val = ${$arg[0]};
+            if (${$arg[0]} !== $arg[2]) {
+                $this->defaultContext[$arg[1]] = $val;
+                @trigger_error(sprintf('The "%s" parameter is deprecated since Symfony 4.2, use the "%s" key of the context instead.', $arg[0], $arg[1]), E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/CsvEncoder.php]: instead of this complex logic, I'd suggest to now deal with mixed signatures: either the 1st arg is an array - or not and then we trigger one deprecation and we build the defaultcontext out of the arguments  (like done in JsonDecode)

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 3]
```diff
@@ -30,42 +30,89 @@ class XmlEncoder implements EncoderInterface, DecoderInterface, NormalizationAwa
...
+        $args = array(
+            array('loadOptions', static::LOAD_OPTIONS,  null),
+            array('decoderIgnoredNodeTypes', static::DECODER_IGNORED_NODE_TYPES, $this->defaultContext[static::DECODER_IGNORED_NODE_TYPES]),
+            array('encoderIgnoredNodeTypes', static::ENCODER_IGNORED_NODE_TYPES, $this->defaultContext[static::ENCODER_IGNORED_NODE_TYPES]),
+        );
+        foreach ($args as $arg) {
+            $val = ${$arg[0]};
+            if (${$arg[0]} !== $arg[2]) {
+                $this->defaultContext[$arg[1]] = $val;
+                @trigger_error(sprintf('The "%s" parameter is deprecated since Symfony 4.2, use the "%s" key of the context instead.', $arg[0], $arg[1]), E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: same note as for CsvEncoder

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 4]
```diff
@@ -517,17 +548,17 @@ private function createDomDocument(array $context): \DOMDocument
...
-            'xml_encoding' => 'encoding',
+            static::ENCODING => 'encoding',
             // whether the document is standalone
-            'xml_standalone' => 'xmlStandalone',
+            static::STANDALONE => 'xmlStandalone',
         );
         foreach ($xmlOptions as $xmlOption => $documentProperty) {
-            if (isset($context[$xmlOption])) {
-                $document->$documentProperty = $context[$xmlOption];
+            if ($contextOption = ($context[$xmlOption] ?? $this->defaultContext[$xmlOption] ?? false)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: no need for the brackets

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 5]
```diff
@@ -32,19 +32,37 @@
...
+    protected $defaultContext = array(
+        self::ALLOW_EXTRA_ATTRIBUTES => true,
+        self::CIRCULAR_REFERENCE_LIMIT => 1,
+        self::IGNORED_ATTRIBUTES => array(),
+    );
+
+    /**
+     * @deprecated since Symfony 4.2
+     *
      * @var int
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: let's drop the type meanwhile (same below :) )

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 6]
```diff
@@ -388,8 +441,8 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
                     // Don't run set for a parameter passed to the constructor
                     $params[] = $parameterData;
                     unset($data[$key]);
-                } elseif (isset($context[static::DEFAULT_CONSTRUCTOR_ARGUMENTS][$class][$key])) {
-                    $params[] = $context[static::DEFAULT_CONSTRUCTOR_ARGUMENTS][$class][$key];
+                } elseif (null !== $param = ($context[static::DEFAULT_CONSTRUCTOR_ARGUMENTS][$class][$key] ?? $this->defaultContext[static::DEFAULT_CONSTRUCTOR_ARGUMENTS][$class][$key] ?? null)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: extra brackets

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 7]
```diff
@@ -399,9 +452,9 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
 
             if ($constructor->isConstructor()) {
                 return $reflectionClass->newInstanceArgs($params);
-            } else {
-                return $constructor->invokeArgs(null, $params);
             }
+
+            return $constructor->invokeArgs(null, $params);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: lets keep the original and safe some merge conflicts

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 8]
```diff
@@ -32,16 +32,19 @@
  */
 abstract class AbstractObjectNormalizer extends AbstractNormalizer
 {
-    const ENABLE_MAX_DEPTH = 'enable_max_depth';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: lets keep the original and safe some merge conflicts

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 9]
```diff
@@ -89,20 +92,25 @@ public function normalize($object, $format = null, array $context = array())
...
             $maxDepthReached = false;
-            if (null !== $attributesMetadata && ($maxDepthReached = $this->isMaxDepthReached($attributesMetadata, $class, $attribute, $context)) && !$this->maxDepthHandler) {
+            if (null !== $attributesMetadata && ($maxDepthReached = $this->isMaxDepthReached($attributesMetadata, $class, $attribute, $context)) && !$maxDepthHandler) {
                 continue;
             }
 
             $attributeValue = $this->getAttributeValue($object, $attribute, $format, $context);
             if ($maxDepthReached) {
-                $attributeValue = \call_user_func($this->maxDepthHandler, $attributeValue, $object, $attribute, $format, $context);
+                $attributeValue = $maxDepthHandler($attributeValue, $object, $attribute, $format, $context);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: lets keep the original and safe some merge conflicts

File: src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php [hunk 10]
```diff
@@ -49,17 +61,17 @@ public function normalize($object, $format = null, array $context = array())
         }
 
         $result = array(
-            'type' => $context['type'] ?? 'https://symfony.com/errors/validation',
-            'title' => $context['title'] ?? 'Validation Failed',
+            'type' => $context[static::TYPE] ?? $this->defaultContext[static::TYPE] ?? 'https://symfony.com/errors/validation',
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php]: self:: instead of static::?
generally we should use self:: when using constants when possible IMHO

File: src/Symfony/Component/Serializer/Normalizer/ObjectToPopulateTrait.php [hunk 11]
```diff
@@ -19,15 +19,14 @@
...
+     * @param string $key they in which to look for the object to populate.
+     *                    Keeps backwards compatibility with `AbstractNormalizer`
      *
      * @return object|null an object if things check out, null otherwise
      */
     protected function extractObjectToPopulate($class, array $context, $key = null)
     {
-        $key = $key ?: 'object_to_populate';
-
+        $key = $key ?? 'object_to_populate';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/ObjectToPopulateTrait.php]: I'd suggest reverting all changes in this file

File: src/Symfony/Component/Serializer/Tests/Encoder/JsonDecodeTest.php [hunk 12]
```diff
@@ -50,7 +50,7 @@ public function decodeProvider()
         $assoc = array('foo' => 'bar');
 
         return array(
-            array('{"foo": "bar"}', $stdClass, array()),
+            //array('{"foo": "bar"}', $stdClass, array()),
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Encoder/JsonDecodeTest.php]: to be cleaned up I suppose?

**Review by nicolas-grekas — COMMENTED** [review]
Makes a lot of sense!

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 13]
```diff
@@ -173,7 +212,7 @@ public function supportsDecoding($format)
      */
     public function setRootNodeName($name)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: could be deprecated also, alongside with getRootNodeName?

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 14]
```diff
@@ -32,20 +32,36 @@
     use ObjectToPopulateTrait;
     use SerializerAwareTrait;
 
-    const CIRCULAR_REFERENCE_LIMIT = 'circular_reference_limit';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: why changing the order? (same above btw)

File: src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php [hunk 15]
```diff
@@ -25,19 +25,37 @@ class DateTimeNormalizer implements NormalizerInterface, DenormalizerInterface,
...
+        );
+
+        if (!\is_array($defaultContext)) {
+            @trigger_error(sprintf('The "format" parameter is deprecated since Symfony 4.2, use the "%s" key of the context instead.', self::FORMAT_KEY), E_USER_DEPRECATED);
+
+            $defaultContext = array(self::FORMAT_KEY => $defaultContext);
+        }
+
+        if (null !== $timezone) {
+            @trigger_error(sprintf('The "timezone" parameter is deprecated since Symfony 4.2, use the "%s" key of the context instead.', self::TIMEZONE_KEY), E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php]: lets merge with previous "if" as done in other cases

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php (hunk 13 above)
Comment (dunglas) [reply src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: I don't why they have been introduced in the first time, but yes.

**Review by nicolas-grekas — APPROVED** [review]
That will be much cleaner! Ready on my side, thanks.

File: src/Symfony/Component/Serializer/Encoder/CsvEncoder.php [hunk 16]
```diff
@@ -30,20 +30,34 @@ class CsvEncoder implements EncoderInterface, DecoderInterface
...
-        $this->delimiter = $delimiter;
-        $this->enclosure = $enclosure;
-        $this->escapeChar = $escapeChar;
-        $this->keySeparator = $keySeparator;
-        $this->escapeFormulas = $escapeFormulas;
+        if (!\is_array($defaultContext)) {
+            @trigger_error('Passing configuration options directly to the constructor is deprecated since Symfony 4.2, use the default context instead.', E_USER_DEPRECATED);
+
+            $defaultContext = array(
+                self::DELIMITER_KEY => $defaultContext,
```
Comment (ostrolucky) [inline src/Symfony/Component/Serializer/Encoder/CsvEncoder.php]: all of these for first argument should use casting, in this case (string)

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 17]
```diff
@@ -30,42 +30,80 @@ class XmlEncoder implements EncoderInterface, DecoderInterface, NormalizationAwa
...
+     * @param array $defaultContext
      */
-    public function __construct(string $rootNodeName = 'response', int $loadOptions = null, array $decoderIgnoredNodeTypes = array(XML_PI_NODE, XML_COMMENT_NODE), array $encoderIgnoredNodeTypes = array())
+    public function __construct($defaultContext = array(), int $loadOptions = null, array $decoderIgnoredNodeTypes = array(XML_PI_NODE, XML_COMMENT_NODE), array $encoderIgnoredNodeTypes = array())
     {
-        $this->rootNodeName = $rootNodeName;
-        $this->loadOptions = null !== $loadOptions ? $loadOptions : LIBXML_NONET | LIBXML_NOBLANKS;
-        $this->decoderIgnoredNodeTypes = $decoderIgnoredNodeTypes;
-        $this->encoderIgnoredNodeTypes = $encoderIgnoredNodeTypes;
+        $this->defaultContext = array(
```
Comment (ostrolucky) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: Not sure why is this whole array not defined as default directly in property definition, like rest

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 18]
```diff
@@ -425,16 +437,16 @@ private function updateData(array $data, string $attribute, $attributeValue, str
...
-            !$context[static::ENABLE_MAX_DEPTH] ||
+            !$enableMaxDepth ||
             !isset($attributesMetadata[$attribute]) ||
             null === $maxDepth = $attributesMetadata[$attribute]->getMaxDepth()
         ) {
             return false;
         }
 
-        $key = sprintf(static::DEPTH_KEY_PATTERN, $class, $attribute);
+        $key = sprintf(self::DEPTH_KEY_PATTERN, $class, $attribute);
```
Comment (ostrolucky) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: This is unrelated change and might break BC

**Review by ostrolucky — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php (hunk 17 above)
Comment (dunglas) [reply src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: Because the default values contain expressions (it's not allowed in properties definitions).

File: src/Symfony/Component/Serializer/CHANGELOG.md [hunk 19]
```diff
@@ -23,6 +24,16 @@ CHANGELOG
...
+ * `AbstractNormalizer::$circularReferenceLimit`, `AbstractNormalizer::$circularReferenceHandler`,
+   `AbstractNormalizer::$callbacks`, `AbstractNormalizer::$ignoredAttributes`,
+   `AbstractNormalizer::$camelizedAttributes`, `AbstractNormalizer::setCircularReferenceLimit()`,
+   `AbstractNormalizer::setCircularReferenceHandler()`, `AbstractNormalizer::setCallbacks()` and
+   `AbstractNormalizer::setIgnoredAttributes()` are deprecated, use the default context instead.
+ * `AbstractObjectNormalizer::$maxDepthHandler` and `AbstractObjectNormalizer::setMaxDepthHandler()`
+   are deprecated, use the default context instead.
+ * passing configuration options directly to the constructor of `CsvEncoder`, `JsonDecode` and
+   `XmlEncoder` is deprecated since Symfony 4.2, use the default context instead.
+>>>>>>> [Serializer] Refactor and uniformize the config by introducing a default context
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/CHANGELOG.md]: extra line here

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php (hunk 18 above)
Comment (dunglas) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: It was actually not working in some cases. It's safer this way and will prevent people from relying on overriding such constants.

File: src/Symfony/Component/Serializer/Encoder/JsonDecode.php [hunk 20]
```diff
@@ -22,19 +22,41 @@ class JsonDecode implements DecoderInterface
 {
     protected $serializer;
 
-    private $associative;
-    private $recursionDepth;
+    /**
+     * True to return the result associative array, false for a nested stdClass hierarchy.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/JsonDecode.php]: missing "as an" before "associative"?

File: src/Symfony/Component/Serializer/Encoder/JsonEncode.php [hunk 21]
```diff
@@ -20,11 +20,24 @@
  */
 class JsonEncode implements EncoderInterface
 {
-    private $options;
+    const JSON_ENCODE_OPTIONS = 'json_encode_options';
 
-    public function __construct(int $bitmask = 0)
+    private $defaultContext = array(
+        'json_encode_options' => 0,
```
[the remark this answers is not in the document]
Comment (dunglas) [reply src/Symfony/Component/Serializer/Encoder/JsonEncode.php]: It looks like it doesn't exist: https://github.com/php/php-src/blob/49a4e695845bf55e059e7f88e54b1111fe284223/ext/json/php_json.h#L61

[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/Serializer/Encoder/JsonEncode.php]: I meant s/'json_encode_options'/self::JSON_ENCODE_OPTIONS/

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 22]
```diff
@@ -38,14 +38,30 @@
...
     const DEFAULT_CONSTRUCTOR_ARGUMENTS = 'default_constructor_arguments';
+    const CALLBACKS = 'callbacks';
+    const CIRCULAR_REFERENCE_HANDLER = 'circular_reference_handler';
+    const IGNORED_ATTRIBUTES = 'ignored_attributes';
 
     /**
-     * @var int
+     * @internal
+     */
+    const CIRCULAR_REFERENCE_LIMIT_COUNTERS = 'circular_reference_limit_counters';
```
[the remark this answers is not in the document]
Comment (dunglas) [reply src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: I added a new mechanism to deal with this in a more generic way: https://github.com/symfony/symfony/pull/28709/files#diff-cd90b97b50abf20fa94468689e7486afR474

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php (hunk 17 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: which one? `LIBXML_NONET | LIBXML_NOBLANKS` works also as it's resolved at compile time if that's what you have in mind

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 23]
```diff
@@ -52,9 +56,10 @@
      */
     protected $classDiscriminatorResolver;
 
-    public function __construct(ClassMetadataFactoryInterface $classMetadataFactory = null, NameConverterInterface $nameConverter = null, PropertyTypeExtractorInterface $propertyTypeExtractor = null, ClassDiscriminatorResolverInterface $classDiscriminatorResolver = null, callable $objectClassResolver = null)
+    public function __construct(ClassMetadataFactoryInterface $classMetadataFactory = null, NameConverterInterface $nameConverter = null, PropertyTypeExtractorInterface $propertyTypeExtractor = null, ClassDiscriminatorResolverInterface $classDiscriminatorResolver = null, callable $objectClassResolver = null, array $defaultContext = array())
     {
-        parent::__construct($classMetadataFactory, $nameConverter);
+        parent::__construct($classMetadataFactory, $nameConverter, $defaultContext);
+        $this->defaultContext[self::EXCLUDE_FROM_CACHE_KEY] = [self::CIRCULAR_REFERENCE_LIMIT_COUNTERS];
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: `s/[]/array()`

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 24]
```diff
@@ -457,6 +471,11 @@ private function isMaxDepthReached(array $attributesMetadata, string $class, str
      */
     private function getCacheKey(?string $format, array $context)
     {
+        foreach($context[self::EXCLUDE_FROM_CACHE_KEY] ?? $this->defaultContext[self::EXCLUDE_FROM_CACHE_KEY] as $key) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: missing space after foreach

**Review by nicolas-grekas — APPROVED** [review]
Double :+1: from me, especially after the last changes.

**Review by ostrolucky — COMMENTED** [review]
Seems great. However, shouldn't we also update UPGRADE document?

**Review by ogizanagi — APPROVED** [review]

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 25]
```diff
@@ -169,21 +206,29 @@ public function supportsDecoding($format)
...
      * Sets the root node name.
      *
+     * @deprecated
+     *
      * @param string $name Root node name
      */
     public function setRootNodeName($name)
     {
-        $this->rootNodeName = $name;
+        @trigger_error(sprintf('The "%s()" method is deprecated since Symfony 4.1, use the context instead.', __METHOD__), E_USER_DEPRECATED);
```
Comment (ro0NL) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: `4.2`

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 26]
```diff
@@ -169,21 +206,29 @@ public function supportsDecoding($format)
...
      * Returns the root node name.
      *
+     * @deprecated
+     *
      * @return string
      */
     public function getRootNodeName()
     {
-        return $this->rootNodeName;
+        @trigger_error(sprintf('The "%s()" method is deprecated since Symfony 4.1, use the context instead.', __METHOD__), E_USER_DEPRECATED);
```
Comment (ro0NL) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: `4.2`

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 27]
```diff
@@ -169,21 +206,29 @@ public function supportsDecoding($format)
...
-        $this->rootNodeName = $name;
+        @trigger_error(sprintf('The "%s()" method is deprecated since Symfony 4.1, use the context instead.', __METHOD__), E_USER_DEPRECATED);
+
+        $this->defaultContext[self::ROOT_NODE_NAME] = $name;
     }
 
     /**
      * Returns the root node name.
      *
+     * @deprecated
```
Comment (ro0NL) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: `@deprecated since Symfony 4.2`

**Comment by nicolas-grekas** [comment]
(rebase needed + some comments to fix)
